contracts/rules/TransferRules.sol

```diff
@@ -56,17 +57,26 @@ contract TransferRules is ITransferRules, ManualApproval, Whitelisted {
     * @param value The amount of tokens to send.
     */
     function doTransfer(address from, address to, uint256 value) external onlySRC20 returns (bool) {
+        require(
+            (isWhitelisted(from) || isGrayListed(from)) &&
+            (isWhitelisted(to) || isGrayListed(to)),
+            "Transfer not authorized"
+        );
+
         if (isWhitelisted(from) && isWhitelisted(to)) {
             if (isGrayListed(from) || isGrayListed(to)) {
                 _transferRequest(from, to, value);
             } else {
                 require(ISRC20(_src20).executeTransfer(from, to, value), "SRC20 transfer failed");
             }
-        } else if (isGrayListed(from) && isWhitelisted(to) ||
+        }
```

Comment:
**Notice**: If `from` and `to` are both in white and gray list, `_transferRequest` will be executed twice?

Can this code (after require that checks are both parties on one of the lists) be:
```
// someone is on gray list
if (isGrayListed(from) || isGrayListed(to)) {
    _transferRequest(from, to, value);
    return true;
}

// no one is on gray list
require(ISRC20(_src20).executeTransfer(from, to, value), "SRC20 transfer failed");
        
return true;
```
?